### PR TITLE
parser: make `text` return utf8 encoding string

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close  https://github.com/pingcap/tidb/issues/31228

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

File: parser/ast/ast.go [hunk 1]
```diff
@@ -16,6 +16,7 @@
 package ast
 
 import (
+	"github.com/pingcap/tidb/parser/charset"
```
Comment (zimulala) [inline parser/ast/ast.go:19]: Please format it.

File: parser/ast/base.go [hunk 2]
```diff
@@ -35,12 +37,22 @@ func (n *node) OriginTextPosition() int {
 }
 
 // SetText implements Node interface.
-func (n *node) SetText(text string) {
+func (n *node) SetText(enc charset.Encoding, text string) {
```
Comment (zimulala) [inline parser/ast/base.go:40]: Please add tests for this function.

File: parser/ast/base.go [hunk 3]
```diff
@@ -35,12 +37,22 @@ func (n *node) OriginTextPosition() int {
...
+	if enc == nil {
+		enc = charset.EncodingBinImpl
+	}
+	utf8Lit, _ := enc.Transform(nil, charset.HackSlice(text), charset.OpDecodeReplace)
```
Comment (Defined2014) [inline parser/ast/base.go:44]: Please add a performance test result for GBK stmt.

Comment (xiongjiwei) [reply parser/ast/base.go:44]: it definitely worse than the utf8 case, I will have the performance with the whole GBK feature

Comment (Defined2014) [reply parser/ast/base.go:44]: IMO, I think this PR will reduce GBK's performance. Becaues we add one more transform for GBK. And I want to know the exact number. Just a simple test with parser only.

Comment (Defined2014) [reply parser/ast/base.go:44]: like this [link](https://github.com/pingcap/tidb/pull/31039)
image

Comment (xiongjiwei) [reply parser/ast/base.go:44]: the same test case
```
BenchmarkUTF8-8   	     637	   1865371 ns/op
BenchmarkGBK-8   	     122	   9589314 ns/op
```

Comment (Defined2014) [reply parser/ast/base.go:44]: Seems declined a lot. 😭

Comment (xiongjiwei) [reply parser/ast/base.go:44]: image

**Review by Defined2014 — APPROVED** [review]

File: cmd/explaintest/t/new_character_set.test [hunk 4]
```diff
@@ -63,3 +63,8 @@ set character_set_connection = utf8mb4;
 create table t(j json, b blob, s1 varchar(255) collate binary, s2 varchar(255), st set('{"点赞": "你好"}'), en enum('{"点赞": "你好"}'));
 insert into t values('{"点赞": "你好"}', '{"点赞": "你好"}', '{"点赞": "你好"}', '{"点赞": "你好"}', '{"点赞": "你好"}', '{"点赞": "你好"}');
 select * from t;
+
+set names utf8mb4;
+set @@character_set_client=gbk;
+set @@character_set_connection=gbk;
+select hex('一a'), '一a';
```
Comment (tangenta) [inline cmd/explaintest/t/new_character_set.test:70]:
```suggestion
select hex('一a'), '一a';

```

File: parser/ast/base.go [hunk 5]
```diff
@@ -35,12 +42,26 @@ func (n *node) OriginTextPosition() int {
...
+	n.utf8Text = text
 	n.text = text
 }
 
 // Text implements Node interface.
 func (n *node) Text() string {
+	n.once.Do(func() {
+		if n.enc == nil {
+			return
+		}
```
Comment (tangenta) [inline parser/ast/base.go:56]: Is it possible to return empty string?

File: parser/ast/base.go [hunk 6]
```diff
@@ -35,12 +42,30 @@ func (n *node) OriginTextPosition() int {
 }
 
 // SetText implements Node interface.
-func (n *node) SetText(text string) {
+func (n *node) SetText(enc charset.Encoding, text string) {
+	n.enc = enc
 	n.text = text
+	n.once = &sync.Once{}
```
Comment (tangenta) [inline parser/ast/base.go:48]: We don't need to initialize `n.once` if the encoding is `nil`.

**Review by tangenta — APPROVED** [review]

**Pull request merged**